### ISSUE TITLE
Broken deserialization of Lists with offsets

### ISSUE THREAD

**Issue by jkylling** [chmp/serde_arrow#248]
We do not correctly deserialize record batches to structs with vectors when the GenericListArrays withing the record batch if the record batch has been sliced.

The following test fails
```rust
    #[test]
    fn record_batch_slice() {
        #[derive(serde::Deserialize, serde::Serialize, Debug, PartialEq)]
        struct Struct {
            string: Vec<String>,
        }
        let data = (1..=10).map(|x| Struct { string: vec![x.to_string()] }).collect::<Vec<_>>();
        let fields = Vec::<FieldRef>::from_type::<Struct>(TracingOptions::default()).unwrap();
        let batch = serde_arrow::to_record_batch(&fields, &data).unwrap();
        let batch = batch.slice(5, 5);

        let actual: Vec<Struct> = serde_arrow::from_record_batch(&batch).unwrap();
        assert_eq!(data[5..10], actual);
    }
```

It might be that we don't account for `GenericListArray::offsets` when reading the vectors.

**Comment by chmp** [comment]
Thanks @jkylling for the report and the reproducing example!

**Comment by chmp** [comment]
You're right. The offset array is assumed to always start at `0` in the current impl. However, `arrow` does not slice the values, but directly the offset array without slicing the values.

**Comment by chmp** [comment]
@jkylling The underlying issue is that current impl uses an iterator-design to access the underlying arrays. This design only allows to get the next value, but does not allow random access. 

The current impl only checks how many elements each list has (`offsets[i + 1] - offsets[i]`) and consumes that many elements without checking where exactly it is located. For sliced lists (and maps) the offset array will start with a non-zero value, e.g., in your example `5`. A quick fix is to consume that many elements up front from the underlying values array. TBH. I don't like this solution, but can't really think of a better way without major changes how the deserialization logic is structured.

**Comment by jkylling** [comment]
The quick fix would come with a performance penalty if batches are processed in smaller sub batches like
```
batch.slice(0, 100)
batch.slice(100, 200)
batch.slice(200, 300)
...
```
Then each element of a vector is processed O(n * number of sub batches) times.

**Comment by chmp** [comment]
Definitely. Unfortunately I expect that a proper solution will take a bit longer to be implemented.

**Comment by chmp** [comment]
Fixed in `0.12.2`. ATM I don't have the capacity to implement a fix without upfront work. I created #250 to follow up on this issue.

Thanks a lot @jkylling  for bringing this issue to my attention.

**Comment by jkylling** [comment]
Thank you for the quick turnaround!

**Comment by chmp** [comment]
I implemented a proper fix with #260. It will be released with 0.13. Thanks prompting me to look into a proper fix. Now the impl is simpler to easier understand and more powerfull :)
